service/controller/resource/clusterupgrade/create.go

```diff
@@ -111,57 +112,70 @@ func (r *Resource) ensureMasterHasUpgraded(ctx context.Context, cluster capiv1al
 }
 
 func (r *Resource) propagateReleaseToMachinePools(ctx context.Context, cr capiv1alpha3.Cluster) error {
-	lst := expcapiv1alpha3.MachinePoolList{}
-	err := r.ctrlClient.List(ctx, &lst, client.MatchingLabels{capiv1alpha3.ClusterLabelName: key.ClusterName(&cr)})
+	lst, err := helpers.GetMachinePoolsByMetadata(ctx, r.ctrlClient, cr.ObjectMeta)
 	if err != nil {
 		return microerror.Mask(err)
 	}
 
-	if !hasMachinePoolsUpgradeCompleted(lst.Items, cr.Labels[label.ReleaseVersion]) {
-		r.logger.LogCtx(ctx, "level", "debug", "message", "machinepool still upgrading; skipping release propagation")
-		return nil
-	}
+	allStartedNodePoolUpgradesAreCompleted := true
+	var machinePoolUpgradeCandidate *capiv1alpha3exp.MachinePool
 
 	for _, mp := range lst.Items {
-		if cr.Labels[label.ReleaseVersion] != mp.Labels[label.ReleaseVersion] ||
-			cr.Labels[label.AzureOperatorVersion] != mp.Labels[label.AzureOperatorVersion] {
-
-			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updating release to machinepool %q", mp.Name))
-
-			mp.Labels[label.ReleaseVersion] = cr.Labels[label.ReleaseVersion]
-			mp.Labels[label.AzureOperatorVersion] = cr.Labels[label.AzureOperatorVersion]
-
-			err = r.ctrlClient.Update(ctx, &mp)
-			if apierrors.IsConflict(err) {
-				r.logger.LogCtx(ctx, "level", "debug", "message", "conflict trying to save object in k8s API concurrently", "stack", microerror.JSON(microerror.Mask(err)))
-				r.logger.LogCtx(ctx, "level", "debug", "message", "cancelling resource")
-				return nil
-			} else if err != nil {
+		machinePool := mp
```

Comment:
Any reason why not just rename the iterator variable?

Reply:
Further down in code `machinePool` pointer is saved to be used outside of `for` loop. I believe there would be issues if using a pointer to the var declared by `for` loop, that issue where the var `machinePoolUpgradeCandidate` outside of func would point to the last element in the slice (but now when I look at it, that I guess would not be the issue here, as we are not iterating all the way to the end).

Anyway, should not matter if this gets refactored and split into smaller components like described in another comment.